### PR TITLE
For #40555: Bundles httplib from Python 2.7.5 source and uses it when needed.

### PR DESCRIPTION
**Pull request by thebeeland**
There is a situation on CentOS 7.x where Houdini comes bundled with an ssl.py that is incompatible with the system Python's httplib. This breaks shotgun_api3 when attempting to connect to Shotgun over https. The fix is to bundle the compatible httplib.py with the engine and force its use when we detect a possible compatibility issues with ssl.

### CODE REVIEW

File: classic_startup/python2.7libs/pythonrc.py [hunk 1]
```diff
@@ -43,6 +44,26 @@ def classic_startup():
     # add to the system path
     sys.path.insert(0, tk_houdini_python_path)
 
+    # We need to check to make sure we don't have an incompatibility
+    # between httplib and Houdini's bundled ssl.py. This is a problem
+    # on some Linux distros (CentOS 7.x) with H16.
+    if sys.platform == "linux2" and sys.version.startswith("2.7.5"):
```
Comment (josh-t) [inline classic_startup/python2.7libs/pythonrc.py]: Recommend 
```python
if sys.platform.startswith("linux") and ...
```
as per: https://docs.python.org/2.7/library/sys.html#sys.platform for eventual compatibility with python `3.3+`.

File: classic_startup/python2.7libs/pythonrc.py [hunk 2]
```diff
@@ -43,6 +44,26 @@ def classic_startup():
...
+        # We can check to see if ssl has the function we know that
+        # system httplib is likely to require. If it doesn't have it,
+        # then we need to force the use of our bundled httplib before
+        # we let the bootstrap happen.
+        import ssl
```
Comment (josh-t) [inline classic_startup/python2.7libs/pythonrc.py]: We're checking `ssl` for a protected method that may be called by `httplib`. Am I understanding that right?

File: classic_startup/python2.7libs/pythonrc.py [hunk 3]
```diff
@@ -43,6 +44,26 @@ def classic_startup():
...
+        if not hasattr(ssl, "_create_default_https_context"):
+            # Clear httplib if it's already been imported.
+            if "httplib" in sys.modules:
+                del sys.modules["httplib"]
+
+            # Add the submodule containing httplib to sys.path so that
+            # the next time it's imported it'll come from there instead
+            # of the system Python install.
+            httplib_path = os.path.join(tk_houdini_python_path, "packages")
+            sys.path.insert(0, httplib_path)
```
Comment (josh-t) [inline classic_startup/python2.7libs/pythonrc.py]: I'm a little worried about deleting from `sys.modules` without replacing it. Would it be better to manually repopulate here to be safe?

File: plugins/basic/python2.7libs/pythonrc.py [hunk 4]
```diff
@@ -32,6 +32,32 @@ def plugin_startup():
     current_dir_path = os.path.dirname(current_file_path)
     plugin_root_path = os.path.dirname(current_dir_path)
 
+    # We need to check to make sure we don't have an incompatibility
+    # between httplib and Houdini's bundled ssl.py. This is a problem
+    # on some Linux distros (CentOS 7.x) with H16.
+    if sys.platform == "linux2" and sys.version.startswith("2.7.5"):
```
Comment (josh-t) [inline plugins/basic/python2.7libs/pythonrc.py]: same comment about `startswith`.

File: plugins/basic/python2.7libs/pythonrc.py [hunk 5]
```diff
@@ -32,6 +32,32 @@ def plugin_startup():
...
+                os.path.dirname(plugins_path),
+                "python",
+                "packages",
+            )
+
+            sys.path.insert(0, packages_path)
+
+            # Clear httplib if it's already been imported.
+            if "httplib" in sys.modules:
+                del sys.modules["httplib"]
```
Comment (josh-t) [inline plugins/basic/python2.7libs/pythonrc.py]: same comment about deleting without replacing.

File: classic_startup/python2.7libs/pythonrc.py (hunk 2 above)
Comment (thebeeland) [inline classic_startup/python2.7libs/pythonrc.py]: Yes, you are. Unfortunately. httplib2 uses httplib, which calls an _ method on ssl.

File: classic_startup/python2.7libs/pythonrc.py (hunk 3 above)
Comment (thebeeland) [inline classic_startup/python2.7libs/pythonrc.py]: I'm now importing httplib in the global scope after bootstrap.

File: plugins/basic/python2.7libs/pythonrc.py (hunk 5 above)
Comment (thebeeland) [inline plugins/basic/python2.7libs/pythonrc.py]: I'm now importing httplib in the global scope after bootstrap.

File: classic_startup/python2.7libs/pythonrc.py [hunk 6]
```diff
@@ -43,6 +44,26 @@ def classic_startup():
     # add to the system path
     sys.path.insert(0, tk_houdini_python_path)
 
+    # We need to check to make sure we don't have an incompatibility
+    # between httplib and Houdini's bundled ssl.py. This is a problem
+    # on some Linux distros (CentOS 7.x) with H16.
```
Comment (robblau) [inline classic_startup/python2.7libs/pythonrc.py]: Can you add a bit about how the setup of python path in houdini will bleed through to the system httplib which for older distros could be incompatible with the libraries from the houdini distribution?  The current comment does good for the what of what is happening, but there is some missing why (which down the road will be good to have when we want to remove this and don't remember why it is here).

**Pull request merged**